### ISSUE TITLE
Implement pre-training tasks for GNNs

### ISSUE THREAD

**Issue by egor-bogomolov** [JetBrains-Research/JetNN#21]
We need to outline the existing pre-training tasks that researchers used to pre-train GNNs and implement the most promising in this repo.

**Comment by petroolg** [comment]
I've done some research on pre-training strategies for GNN. Here I list some recent methods. 
But before there are two observations:
1) everybody is doing an out-of-fold evaluation, arguing that it is necessary when working with graphs,
2) in some cases GNN pre-training leads to negative transfer when pre-training harms final performance. 

The first issue applies to us as well, our graph distribution is not static. It changes through time and depending on the domain (application of code). To avoid negative transfer we need to always test on the final task (maybe this is obvious).  

https://cs.stanford.edu/people/jure/pubs/pretrain-iclr20.pdf
**Hu at al. STRATEGIES FOR PRE-TRAINING GRAPH NEURAL NETWORKS (ICLR 2020)**
The authors outline two directions:

- Node-level 
   - context prediction
   - masking node and edge attributes
- Graph-level
  - Supervised graph-level property prediction

They emphasize that multi-task pre-training is necessary to avoid negative transfer. 
They also use a particular splitting strategy to evaluate out-of-distribution performance. 

https://arxiv.org/pdf/2006.15437.pdf
**Hu at al. GPT-GNN: Generative Pre-Training of Graph Neural Networks 2020**
Authors factorize the likelihood of graph generation into two components: 1) Attribute Generation and 2) Edge Generation.
They compare themselves with 3 baselines (which might themselves be promising directions):
- GAE, which denotes graph auto-encoders, focuses on a traditional link prediction task. It randomly masks out a fixed proportion of the edges and asks the model to reconstruct these masked edges.
- GraphSAGE forces connected nodes to have similar output node embeddings. Its main difference from GAE lies in that it does not mask out the edges during pre-training. 
- Graph Infomax tries to maximize the local node embeddings with global graph summary embeddings. Following its setting for a large-scale graph, for each sampled subgraph, we shuffle the graph to construct negative samples

The authors test the models' performance on 3 graph transfer settings:
- time transfer,
- field transfer,
- time + field transfer,

basically testing out-of-fold performance.

Another source of knowledge worth mentioning is a talk by Prof. Chuan Shi **Large-Scale Graph Pre-training** http://www.shichuan.org/doc/graph%20pretrain.pdf which mentions the above approaches. But I have not yet studied it thoroughly.

**Comment by egor-bogomolov** [comment]
Thoughts from (4) are applicable in our case: the set of node types for intermediate nodes (non-tokens) is rather small compared to the token vocabulary. Let's also look in this direction! Yet, for the nodes with tokens vocabulary is large and is more NLP-like.

For graph-level tasks, we can generate documentation when it is available, wdyt?

**Comment by petroolg** [comment]
Yeah, sounds great, documentation generation is a killer task on graph level. I'd add generation token names where possible on the node level and multi-label classification on the graph level predicting which libraries were used (if this info is not already contained in nodes).
So, I'm starting with Contextual Property Prediction from (4). But I also need a "reference task" to check if pre-training harms or helps. I see two use cases directly inside the package: 1) variable misuse and 2) variable naming. Let's pick one of them for start.
